pkg/trace/api/container_linux.go

```diff
@@ -0,0 +1,187 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022-present Datadog, Inc.
+
+package api
+
+import (
+	"context"
+	"github.com/DataDog/datadog-agent/pkg/util/cgroups"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"syscall"
+	"time"
+)
+
+type ucredKey struct{}
+
+// connContext injects a Unix Domain Socket's User Credentials into the
+// context.Context object provided. This is useful as the connContext member of an http.Server, to
+// provide User Credentials to HTTP handlers.
+//
+// If the connection c is not a *net.UnixConn, the unchanged context is returned.
+func connContext(ctx context.Context, c net.Conn) context.Context {
+	s, ok := c.(*net.UnixConn)
+	if !ok {
+		return ctx
+	}
+	file, err := s.File()
+	if err != nil {
+		log.Debugf("Failed to obtain unix socket file: %v", err)
+		return ctx
+	}
+	fd := int(file.Fd())
+	ucred, err := syscall.GetsockoptUcred(fd, syscall.SOL_SOCKET, syscall.SO_PEERCRED)
+	if err != nil {
+		log.Debugf("Failed to read credentials from unix socket: %v", err)
+		return ctx
+	}
+	return context.WithValue(ctx, ucredKey{}, ucred)
+}
+
+// cacheExpiration determines how long a pid->container ID mapping is considered valid. This value is
+// somewhat arbitrarily chosen, but just needs to be large enough to reduce latency and I/O load
+// caused by frequently reading mappings, and small enough that pid-reuse doesn't cause mismatching
+// of pids with container ids. A one minute cache means the latency and I/O should be low, and
+// there would have to be thousands of containers spawned and dying per second to cause a mismatch.
+const cacheExpiration = time.Minute
+
+// IDProvider implementations are able to look up a container ID given a ctx and http header.
+type IDProvider interface {
+	GetContainerID(context.Context, http.Header) string
+}
+
+// noCgroupsProvider is a fallback IDProvider that only looks in the http header for a container ID.
+type noCgroupsProvider struct{}
+
+func (i *noCgroupsProvider) GetContainerID(_ context.Context, h http.Header) string {
+	return h.Get(headerContainerID)
+}
+
+// NewIDProvider initializes an IDProvider instance using the provided procRoot to perform cgroups lookups in linux environments.
+func NewIDProvider(procRoot string) IDProvider {
+	reader, err := cgroups.NewReader()
+	if err != nil {
+		log.Warnf("Failed to identify cgroups version due to err: %v. APM data may be missing containerIDs.", err)
+		return &noCgroupsProvider{}
+	}
+	cgroupController := ""
+	if reader.CgroupVersion() == 1 {
+		cgroupController = "memory"
+	}
```

Comment:
Do we understand what this means? It's not obvious to me.
I know a lot of this is copied, but we should understand it.

Reply:
Unfortunately I do not :( It seems to be a sort of important constant that I saw defined here (https://github.com/DataDog/datadog-agent/blob/main/pkg/util/containers/v2/metrics/system/collector_linux.go#L26) but can't import because of where it is. It looks like it defines `where` in the filesystem it should look for cgroup info files and the `memory` ends up as part of the path when using Cgroups v1. Any ideas on who else we could tag here to make sure we're understanding this correctly?

Reply:
@ahmed-mez might know :smile:

Reply:
> It looks like it defines where in the filesystem it should look for cgroup info files and the memory ends up as part of the path when using Cgroups v1.

This is correct, the `memory` controller is used by the cgroupv1 utils in the agent to parse the procfs. I'd say it's an implementation detail.

Reply:
Do we want to make a note here? Magic values like this make code hard to maintain.